turbine/src/cluster_nodes.rs

```diff
@@ -220,30 +227,32 @@ impl ClusterNodes<RetransmitStage> {
         fanout: usize,
         socket_addr_space: &SocketAddrSpace,
     ) -> Result<(/*root_distance:*/ usize, Vec<SocketAddr>), Error> {
-        let mut weighted_shuffle = self.weighted_shuffle.clone();
         // Exclude slot leader from list of nodes.
         if slot_leader == &self.pubkey {
             return Err(Error::Loopback {
                 leader: *slot_leader,
                 shred: *shred,
             });
         }
-        if let Some(index) = self.index.get(slot_leader) {
-            weighted_shuffle.remove_index(*index);
-        }
-        let mut rng = get_seeded_rng(slot_leader, shred);
-        let (index, peers) = get_retransmit_peers(
-            fanout,
-            |k| self.nodes[k].pubkey() == &self.pubkey,
-            weighted_shuffle.shuffle(&mut rng),
-        );
-        let protocol = get_broadcast_protocol(shred);
-        let peers = peers
-            .filter_map(|k| self.nodes[k].contact_info()?.tvu(protocol))
-            .filter(|addr| socket_addr_space.check(addr))
-            .collect();
-        let root_distance = get_root_distance(index, fanout);
-        Ok((root_distance, peers))
+        THREAD_LOCAL_WEIGHTED_SHUFFLE.with_borrow_mut(|weighted_shuffle| {
```

Comment:
This will panic if borrowed, and the function takes &self, so presumably can be called from multiple threads? There should be at least a comment somewhere that makes this invariant explicit (assuming it's always held)

Reply:
> can be called from multiple threads

as @alexpyattaev pointed out on discord, this is a dumb pre-coffee comment because it's a thread local. The code in the closure accesses self but it's unlikely that it'll ever re-enter, so I think we're fine